File: exploration/crawler.py
import requests
import re
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
from config import BASE, AUTH

# Captures the target secret format; reused across body, headers, and raw bytes.
PATTERN = re.compile(r"VISUALPING\{[0-9a-fA-F]{16}\}")

# Hard cap so a query-string explosion can never make the crawl run forever.
# Bump this (1000, 2000, ...) to crawl deeper — more pages = more passwords.
MAX_URLS = 1000


def same_host(url):
    """Only follow links that stay on the target host."""
    # Compare host:port only, so http/https and path differences don't matter.
    return urlparse(url).netloc == urlparse(BASE).netloc


def normalize_url(url):
    """Strip query string and fragment so ?v=6 and ?utm_source=... are the
    same resource — otherwise each variant inflates the queue and we never
    reach the pages that actually contain passwords."""
    parsed = urlparse(url)
    return parsed._replace(query="", fragment="").geturl()


def get_all_urls(html, base_url):
    soup = BeautifulSoup(html, "html.parser")
    urls = set()
    tags = [("a", "href"), ("script", "src"), ("link", "href"),
            ("img", "src"), ("iframe", "src"), ("source", "src"),
            ("form", "action")]
    for tag_name, attr in tags:
        for tag in soup.find_all(tag_name):
            value = tag.get(attr)
            if value:
                urls.add(urljoin(base_url, value))
    return urls


def crawl():
    visited = set()
    queue = [BASE]
    found_passwords = set()

    # Breadth-first traversal: pop the oldest item, so we fully cover the
    # homepage's direct children before descending deeper.
    while queue:
        url = queue.pop(0)
        # Normalize before dedup so query-string variants count as the same page.
        url = normalize_url(url)
        # Skip anything already seen — prevents re-fetching and cycles.
        if url in visited:
            continue
        visited.add(url)
        print(f"Fetching: {url}")

        # Bounded crawl: stop once we've fetched enough URLs, so a query-string
        # explosion can never make this run forever.
        if len(visited) >= MAX_URLS:
            print(f"\nReached the {MAX_URLS}-URL cap — stopping the crawl.")
            break

        try:
            resp = requests.get(url, auth=AUTH, timeout=10)
        except Exception as e:
            # A single failed URL shouldn't abort the whole crawl.
            print(f"  ERROR: {e}")
            continue

        # Search body, headers, and raw bytes
        found_passwords.update(PATTERN.findall(resp.text))
        found_passwords.update(PATTERN.findall(str(resp.headers)))
        found_passwords.update(
            PATTERN.findall(resp.content.decode("utf-8", errors="ignore"))
        )

        # Only parse HTML/CSS/JS for more links
        content_type = resp.headers.get("Content-Type", "")
        if ("text/html" in content_type or "text/css" in content_type
                or "javascript" in content_type):
            for link in get_all_urls(resp.text, url):
                # Normalize before checking the visited set, so a link like
                # /page?ref=related is treated as the same page as /page.
                link = normalize_url(link)
                if same_host(link) and link not in visited:
                    queue.append(link)

    print(f"\nVisited {len(visited)} URLs")
    print(f"Passwords found ({len(found_passwords)}):")
    for p in found_passwords:
        print(" ", p)


if __name__ == "__main__":
    crawl()

File: crawler.js
const { chromium } = require('playwright');
const { BASE, USERNAME, PASSWORD } = require('./config.js');

// Secret shape we hunt for. The /g flag makes .match() return every occurrence.
const PATTERN = /VISUALPING\{[0-9a-fA-F]{16}\}/g;
const MAX_URLS = 600; // safety cap — stop after this many unique URLs

(async () => {
  const browser = await chromium.launch({ headless: true });
  // Context isolates cookies/storage; httpCredentials auto-applies Basic Auth
  // to every request, so we don't have to construct the header ourselves.
  const context = await browser.newContext({
    httpCredentials: { username: USERNAME, password: PASSWORD },
  });

  const visited = new Set();
  const queue = [BASE];
  const foundPasswords = new Set(); // de-dupe — report each password once

  // Listen for responses across all pages in this context.
  // Attaching to `context` (not a single page) means the listener covers every
  // tab we open in the BFS loop below.
  context.on('response', async (response) => {
    let body = '';
    try {
      body = await response.text();
    } catch (e) {
      // Binary file — not text-readable, ignore.
    }

    // Search body + headers together; headers can carry secrets too.
    const allText = body + JSON.stringify(response.headers());
    const matches = allText.match(PATTERN);
    if (matches) {
      matches.forEach((m) => {
        if (!foundPasswords.has(m)) {
          foundPasswords.add(m);
          console.log(`✅ FOUND: ${m}  (in ${response.url()})`);
        }
      });
    }
  });

  // Breadth-first traversal: shift() removes the oldest item, so we fully
  // cover a page's direct children before descending deeper.
  while (queue.length > 0) {
    const url = queue.shift();
    // Skip anything already seen — prevents re-fetching and cycles.
    if (visited.has(url)) continue;
    visited.add(url);

    // Safety cap — stop the crawl once we've visited the limit.
    if (visited.size >= MAX_URLS) {
      console.log(`\n⚠️ Hit MAX_URLS (${MAX_URLS}) — stopping.`);
      break;
    }

    // Stay on the target host — don't crawl out to other domains.
    try {
      const host = new URL(url).hostname;
      const baseHost = new URL(BASE).hostname;
      if (host !== baseHost) continue;
    } catch (e) {
      continue; // malformed URL — ignore it
    }

    console.log(`Visiting: ${url}`);
    try {
      const page = await context.newPage();
      await page.goto(url, { waitUntil: 'networkidle', timeout: 15000 });

      // Collect links from the page DOM.
      // page.evaluate() runs inside the browser, where the browser resolves
      // relative hrefs to absolute URLs automatically.
      const links = await page.evaluate(() => {
        const nodes = Array.from(
          document.querySelectorAll('a[href], script[src], link[href], img[src]')
        );
        return nodes.map((el) => el.href || el.src).filter(Boolean);
      });

      links.forEach((link) => {
        if (!visited.has(link)) queue.push(link);
      });

      await page.close(); // free the tab — BFS can open many
    } catch (e) {
      console.log(`  Error on ${url}: ${e.message}`);
    }
  }

  console.log(`\nVisited: ${visited.size} URLs`);
  console.log(`Found ${foundPasswords.size} passwords:`);
  foundPasswords.forEach((p) => console.log(' ', p));

  await browser.close();
})();